Added a semantic/create test for auto inc attributes.

marnusw opened the pull request:
This is a test for the auto-inc feature of attributes as suggested by @particlebanana [here](https://github.com/balderdashy/sails-memory/pull/20#issuecomment-99590001). When a record is inserted with a value specified and it is larger than the current auto-inc counter the counter should fast-forward to match the inserted value. This happens internally in real databases, but was not working in `sails-memory`.

At the moment this test fails the Travis build due to the bug in `sails-memory`, but it will pass once balderdashy/sails-memory#20 is merged.

marnusw commented:
Thinking about this again, auto-inc is not really [MongoDb friendly](http://docs.mongodb.org/manual/tutorial/create-an-auto-incrementing-field/#considerations). Perhaps this is another test that should wait for #64.

dmarcelino commented:
Thanks @marnusw!

> Thinking about this again, auto-inc is not really MongoDb friendly. Perhaps this is another test that should wait for #64.

That's true Indeed!

devinivy commented:
I don't know that this behavior should be tested in the general-purpose adapter tests.  There are many ways that auto-incrementing could work.  Some (non-waterline-official) adapters also use the auto-incrementing property to make a UUID field on databases that don't support auto-incrementing.

marnusw commented:
@devinivy but even if the auto-increment function is implemented in the adapter level rather than the underlying persistence mechanism wouldn't it still make sense to expect the operation as tested here? It seems like the `autoIncrement` attribute property should specify one and only one type of behavior.

That said, PR #64 suggests specifying a list of "features" in the `package.json` of an adapter to indicate which specialized attribute properties it supports which should thus be tested. Adapters that don't support a feature (for example MongoDb in the case of auto inc) would then not specify `autoIncrement` skip the relevant test entirely.

devinivy commented:
I don't generally agree that auto-increment should have one behavior, but I could be off-base there.  The achieve the same effect, why couldn't some database/adapter increment by 2s instead of 1s?  More realistically, why couldn't a database/adapter implement something like [snowflake](https://blog.twitter.com/2010/announcing-snowflake) for auto-incrementing?  The sails-dynamodb adapter uses the setting to generate UUIDs, since an auto-increment counter wouldn't make much sense for that type of database.  It's more about auto-generating ids.  Waterline was built with SQL in mind originally, so we call it `autoIncrement`.

I could be off-base here!  Maybe @tjwebb or @dmarcelino have thoughts.

In any case, we could add this test directly onto the adapters that are expected to auto-increment in this manner.

dmarcelino commented:
I believe the problem is that the keyword `autoIncrement` is used too broadly. The test proposed by @marnusw seems to cover the default (for some adapters) case where implicitly:
``` javascript
id: {
  type: "integer",
  autoIncrement: true,
  primaryKey: true
}
```

But `autoIncrement` can also be applied to:
* different data types;
* non primaryKey columns.

So @devinivy has a good point. In other words, an adapter can implement `autoIncrement` and still break these tests. So, the way these tests are written are probably not broad enough for all adapters that implement some sort of `autoIncrement`...

We probably should discuss what `autoIncrement` (in the context of waterline) means and what is intended functionality for it. As it stands in current waterline `autoIncrement` functionality the only thing we can *guarantee* is that the adapter won't re-use an existing `id`.

marnusw commented:
> As it stands in current waterline autoIncrement functionality the only thing we can guarantee is that the adapter won't re-use an existing id.

I can rewrite the test to be more general but still ensure the above. It actually describes the problem I encountered in `sails-memory` perfectly and would test that the issue has been resolved. 

The very specialised nature of this test was because I aimed it at the SQL-like `sails-memory` and also because I interpreted the stricter meaning of `autoIncrement`. I guess it's not necessary to be that specific even for `sails-memory` though. Thanks for the feedback!

> We probably should discuss what autoIncrement (in the context of waterline) means and what is intended functionality for it.

If you do formalise this and want me to write the tests for your definition I'd be more than happy to.

devinivy commented:
Thanks a bunch @marnusw – if you have any suggestions for formalizing this we'd love the help!

dmarcelino commented:
@marnusw, PR #64 has been merged, do you mind refactoring this as a "feature"? Thanks!

marnusw commented:
Will do.

marnusw commented:
@devinivy and @dmarcelino, I've been thinking about a formal definition for the `autoIncrement` attribute option and I've come up with the following:

> *Basic Definition:* When `autoIncrement` is set to `true` on an attribute and no value is provided for it a new unique value will be assigned by the adapter before the record is created. It is guaranteed that the adapter will assign a unique value not present on any existing record. The values assigned automatically will not necessarily be sequential, which accommodates the use of UUIDs. If a value for the attribute is present in the data provided for a new record it will be saved as-is without any guarantee of uniqueness. The `autoIncrement` option has no effect when updating existing records. The feature flag is `autoIncrement`.

> *Extended Sequential Definition:* If the adapter will provide sequential unique values, for example increasing integers, then it is further guaranteed that the next value will be the last saved value plus one increment. If a value is provided and is larger than the current auto-inc counter the counter will be bumped to the provided value. If a provided value is less than or equal to the auto-inc counter value the counter will remain unchanged, and again there will be no guaranteed that this value is unique. This extended feature is indicated with the `autoIncrementSequential` feature flag.

I have written a test for the Basic Definition above and pushed it on this PR. The Travis build passes as expected since the feature isn't enabled for any of the adapters. When it is enabled for `sails-memory` it fails, which is expected until balderdashy/sails-memory#20 is merged. The `sails-disk` adapter passes with this feature test enabled. I couldn't test the others right now since I don't have the databases installed on my machine at the moment.

I'll continue to implement a test for the Extended Definition (which I had pretty much done earlier). In the mean time I welcome your feedback.

You'll notice that I'm starting off with a new empty database/adapter before each test. This is to force auto-inc counters to reset. Since the `id` can be an integer or string depending on the adapter it's also quite hard to generate known values for initial records. By using the values generated on the first test as the preset values on the second test this problem is resolved.

marnusw commented:
Tests for sequential auto-inc values have been added, so this PR is complete pending feedback. Once we finalize this and it gets merged I'll update my PR on `sails-memory` to add the two features to its `package.json`.

dmarcelino commented:
@marnusw, I finally got a chance to look deeper at this. I quite like the `autoIncrement.test.js` tests. Have you had the chance to test this against any of the SQL adapters?

I'm not so sure about an `autoIncrementSequential` feature, I think it would most likely belong in the adapters themselves, my reasons are:
* from an API perspective the `autoIncrement` flag is a boolean so the user never gets a chance to choose if he wants a sequential (or whatever) auto increment;
* I'd like to avoid an explosion of features to keep them meaningful and, more or less, obvious.

Anyway let's here what others have to say. When I get the chance I'll pull this to my machine and test against the adapters I have installed.

marnusw commented:
@dmarcelino, thanks for the feedback. OK, I agree that having an explicit `autoIncrementSequential` feature is not ideal. 

I still kind of feel that I'd like to have sequential increments tested when appropriate, which I'd imagine is whenever an integer value is used, but I guess there might just pop up a case where an adapter doesn't want this for some good reason and then the tests are stricter than necessary. 

I'll leave it for comment until the weekend, then I'll take that feature out again. I'll also see if I can make time to run these tests against all the main adapters then. Also for #63.

dmarcelino commented:
> and then the tests are stricter than necessary.

That's my main fear...

> I still kind of feel that I'd like to have sequential increments tested when appropriate

I agree, I'm just unsure if it belongs on waterline-adapter-tests or in the adapter itself.

BTW, if we are going to have features and... *sub-features* I propose we use dot "." notation. Example: `autoIncrement.sequential`. In terms of file structure we could do:

```
features
├- autoIncrement
      ├- core
      ├- sequential 
```

* For an adapter implementing `autoIncrement` only the tests in the `core` folder would run.
* For an adapter implementing `autoIncrement.sequential` both the core and sequential folders would be used.

We would have to update `crossAdapter` and move all contents to `crossAdapter/core`.

marnusw commented:
The "." notation makes a lot of sense to me! It makes the intended outcome much more apparent, and doesn't introduce additional features which are really only special cases. So I'd say definitely the way to go if the consensus is to allow sub-features.

particlebanana commented:
This looks pretty good. I like the dot approach, hopefully it doesn't get too out of hand. All the core adapters should support the sequential feature. Do we want the file structure to look like what @dmarcelino defined above using core as the main test?

marnusw commented:
Two things are still necessary to get this merged:

1. Run the new tests against all core adapters.
2. Make a decision on the "dot approach" and whether to keep `autoIncrement.sequential`

Status:

1. Done: See below and #81.
2. Maybe @devinivy can give the deciding vote?

The results of testing against the various adapters are below. I didn't realize MySQL only supports one autoInc field per table, so I've adapted the tests to account for that. It now only looks for properly incrementing ids.
```
 ------------------------------------------------------------------
| adapter          | version | result | failed | total | wl-sequel |
|------------------|---------|--------|--------|-------|-----------|
| sails-memory     | c7b0610 | failed |      3 |   288 |           |
| sails-disk       | e95628b | passed |      0 |   288 |           |
| sails-mongo      | 021ec02 | passed |      0 |   285 |           |
| sails-mysql      | f6e27d4 | passed |      0 |   313 |     0.4.0 |
| sails-redis      | 72ced91 | failed |      1 |   219 |           |
 -------------------------------------------------------------------
```
I couldn't get PostgreSQL working in the VM yet as mentioned in #81, but the rest have all been run. Since MySQL passes I'm guessing we're OK with Postgres as well, but it's your call. Tested features:

* `sails-memory`: `autoIncrement.sequential` - The errors are expected (balderdashy/sails-memory#20)
* `sails-disk`: `autoIncrement.sequential`
* `sails-mongo`: `autoIncrement` - Sequential increments not supported as far as I know.
* `sails-mysql`: `autoIncrement.sequential`
* `sails-redis`: `autoIncrement.sequential` - Basic autoInc passes, but not sequential. Not sure if it should be supported though.

devinivy commented:
I dig the dot notation and its proposed usage here– nice!

dmarcelino commented:
I guess it's decided, dot '.' notation it is! :+1:

marnusw commented:
Thanks guys, I'll make the changes.

marnusw commented:
I think we're good to go here now. There was actually no need to move the `crossAdapter` test into a "core" folder since it's treated as a special case, but I did so anyway for consistency.

Let me know if you spot anything we still need that I might have missed!

marnusw commented:
@dmarcelino Here is the latest update now that all adapters can be tested with the Vagrant VM, I updated all adapters to the most recent versions:

```
 -----------------------------------
| core modules            | version |
|-------------------------|---------|
| waterline               | 0.10.24 |
| - anchor                |  0.10.3 |
| - waterline-schema      |  0.1.17 |
| waterline-adapter-tests | 7cdff89 |
 -----------------------------------


 ------------------------------------------------------------------
| adapter          | version | result | failed | total | wl-sequel |
|------------------|---------|--------|--------|-------|-----------|
| sails-postgresql | 30317cd | passed |      0 |   342 |     0.4.0 |
| sails-memory     | c7b0610 | failed |      3 |   320 |           |
| sails-disk       | e95628b | passed |      0 |   320 |           |
| sails-mongo      | 021ec02 | passed |      0 |   317 |           |
| sails-mysql      | 01c1a0e | passed |      0 |   345 |     0.4.0 |
| sails-redis      | 72ced91 | passed |      0 |   317 |           |
 -------------------------------------------------------------------
```

### autoIncrement.sequential
* `sails-memory` (Only fails due to balderdashy/sails-memory#20)
* `sails-disk`
* `sails-mysql`

### autoIncrement
* `sails-postgresql`
* `sails-mongodb`
* `sails-redis`

## PostgreSQL and autoIncrement.sequential

It seems that PostgreSQL doesn't support sequential auto increments. Everything works but for the last test, which checks that if a large value is set explicitly, which moves the auto inc counter forward, and then a smaller value is set again the auto inc counter shouldn't be moved back to that value. If this happens and auto incs continue there will eventually be a unique constraint exception, which is the case here!

This seems to be an [implementation detail of PostgreSQL](http://www.postgresql.org/docs/9.3/static/datatype-numeric.html#DATATYPE-SERIAL) and not a problem with the test itself. It seems strange that PostgreSQL would have this problem though. @particlebanana, do you have any insights?

## "." notation implementation

Please note that this was also implemented on this branch in `lib/index.js`, so let me know if you have comments on the way I did that.


**In my opinion this branch is ready to merge!** :smile:

devinivy commented:
Looking good :+1: 
@dmarcelino if you like it, I'd say feel free to push the button!  I think this is technically a blocker for https://github.com/balderdashy/sails-memory/pull/20 right now?

dmarcelino commented:
Thanks @marnusw! Apologies for taking so long but I have been more busy than usual lately...
